src/Aspire.Hosting/Dashboard/proto/resource_service.proto

```diff
@@ -0,0 +1,170 @@
+    // If "display_name" is omitted, this value will be used in UIs.
+    string unique_name = 1;
+
+    // Display string for references to this type in UI. May be localized.
+    // If this value is omitted, UIs will show "unique_name" instead.
+    optional string display_name = 2;
+
+    // Any commands that may be executed against resources of this type, avoiding
+    // the need to copy the value to every Resource instance.
+    //
+    // Note that these commands must apply to matching resources at any time.
+    //
+    // If the set of commands changes over time, use the "commands" property
+    // of the Resource itself.
+    repeated ResourceCommandRequest commands = 3;
+}
+
+////////////////////////////////////////////
+
+message EnvironmentVariable {
+    string name = 1;
+    optional string value = 2;
+    bool is_from_spec = 3;
+}
+
+message Endpoint {
+    string name = 1;
+    optional string allocated_address = 2;
+    optional int32 allocated_port = 3;
+    optional string http_address = 4;
+}
+
+message StringArray {
+    repeated string values = 1;
+}
+
+message AdditionalData {
+    // Name of the data item, e.g. "id", "image", "path", ...
+    string name = 1;
+    // Optional namespace, e.g. "container", "executable", "project", ...
+    optional string namespace = 2;
+    // Optional display name, may be localized
+    optional string display_name = 3;
+    // A single value will be most common. Also support lists, to avoid escaping.
+    oneof kind {
+        string value = 4;
+        StringArray values = 5;
+    }
+}
+
+message ResourceId {
+    string uid = 1;
+    // TODO do we need resource_type to make unique names? if not, inline ResourceId type as string.
+    string resource_type = 2;
+}
+
+// Models the full state of an resource (container, executable, project, etc) at a particular point in time.
+message ResourceSnapshot {
+    ResourceId resource_id = 1;
+    string display_name = 2;
+    optional string state = 3;
+    optional google.protobuf.Timestamp created_at = 4;
+    repeated EnvironmentVariable environment = 5;
+    optional int32 expected_endpoints_count = 6;
+    repeated Endpoint endpoints = 7;
+    repeated ResourceCommandRequest commands = 8;
+
+    // TODO potentially remove endpoints/env vars/anything that might not be general purpose
+
+    // List of additional data, as name/value pairs.
+    // For:
+    // - Containers: image, container_id, ports
+    // - Executables: process_id, executable_path, working_directory, arguments
+    // - Projects: process_id, project_path
+    repeated AdditionalData additional_data = 9;
+}
+
+////////////////////////////////////////////
+
+// Models a snapshot of resource state
+message WatchResourcesSnapshot {
+    repeated ResourceSnapshot resources = 1;
+    repeated ResourceType types = 2;
+}
+
+////////////////////////////////////////////
+
+message ResourceDeletion {
+    ResourceId resource_id = 1;
```

Comment:
So deletion is one-at-a-time, but inserts/updates can be batched? Why the difference?

Reply:
No, both deletions and upserts are multiples.

```
message ResourceDeletion {
    ResourceId resource_id = 1;
}

message WatchResourcesChange  {
    oneof kind {
        ResourceDeletion delete = 1;
        ResourceSnapshot upsert = 2;
    }
}

message WatchResourcesChanges {
    repeated WatchResourcesChange value = 1;
}
```

`WatchResourcesChanges` contains multiple `WatchResourcesChange`, and each of those may be either a delete or upsert.

Note that `ResourceSnapshot` is an immutable snapshot of a single resource, at a point in time. The one scenario where we send multiple resource snapshots is the initial sync, which is modelled as `WatchResourcesSnapshot`.